packages/mobile/src/utils/reachability.ts

```diff
@@ -0,0 +1,81 @@
+import { reachabilityActions } from '@audius/common'
+import type { NetInfoState } from '@react-native-community/netinfo'
+import NetInfo from '@react-native-community/netinfo'
+import { debounce } from 'lodash'
+import { AppState } from 'react-native'
+
+import { env } from 'app/services/env'
+import { dispatch } from 'app/store'
+
+const REACHABILITY_URL = env.REACHABILITY_URL
+
+// Latest connectivity value
+export const Connectivity: { netInfo: NetInfoState | null } = { netInfo: null }
+
+export const checkNetInfoReachability = (netInfo: NetInfoState | null) => {
+  if (!netInfo) return true
+
+  const { isInternetReachable } = netInfo
+  return !!isInternetReachable
+}
+
+// Check that a response from REACHABILITY_URL is valid
+const isResponseValid = (response: Response | undefined) =>
+  response && response.ok
+
+const pingTest = async () => {
+  // If there's no reachability url available, consider ourselves reachable
+  if (!REACHABILITY_URL) {
+    console.warn('No reachability url provided')
+    return true
+  }
+
+  try {
+    const response = await fetch(REACHABILITY_URL, { method: 'GET' })
+
+    if (isResponseValid(response)) {
+      console.debug('Reachability call succeeded')
+      return true
+    }
+    console.debug('Reachability call failed')
+    return false
+  } catch {
+    console.debug('Reachability call failed')
+    return false
+  }
+}
+
+const updateReachability = async (netInfoState: NetInfoState) => {
+  Connectivity.netInfo = netInfoState
+
+  const newValue = checkNetInfoReachability(netInfoState)
+  if (!newValue) {
+    // Don't trust offline signal while app is not in focus
+    // We check on refocus
+    const appState = AppState.currentState
+    if (appState !== 'active') return
+
+    // Perform our own reachability test to be extra sure we're offline
+    const reachable = await pingTest()
+    if (!reachable) {
+      dispatch(reachabilityActions.setUnreachable())
+    }
+  } else {
+    dispatch(reachabilityActions.setReachable())
+  }
+}
+
+export const forceRefreshConnectivity = async () => {
+  NetInfo.refresh()
+  const updatedNetInfoState = await NetInfo.fetch()
+  updateReachability(updatedNetInfoState)
+}
+
+/** Called on first app render */
+export const subscribeToNetworkStatusUpdates = () => {
+  NetInfo.addEventListener(
```

Comment:
is this debounce actually hurting us? i wonder if it's debouncing the "back online" bit?

Reply:
It would definitely also do that. I could maybe separate the debounce into just the part that handles going offline. I think we probably still want that to get delayed.

Reply:
yeah i think that'd be a good call!

going from online => offline should be like pushing a rock up a hill and offline => online should be like letting it roll back down